docs/development.md

```diff
@@ -0,0 +1,775 @@
+          async executionHandler (
+            executionContext: IntegrationStepExecutionContext,
+          ) {
+            return fetchAccounts(executionContext);
+          },
+        },
+      ],
+    },
+
+  ],
+};
+```
+
+### `IntegrationInvocationConfig` fields
+
+#### `instanceConfigFields`
+
+The `instanceConfigFields` field contains a map of integration fields that are
+required to be provided for authenticating with a given provider API. This
+varies between services.
+
+#### `invocationValidator`
+
+The `invocationValidator` field is a validation function that is required for
+ensuring the integration has received a valid set of `instanceConfigFields`. It
+is common practice to execute a test API call to ensure everything has been
+configured properly.
+
+#### `getStepStartStates`
+
+The `getStepStartStates` is an optional function that can be provided for
+determining if a certain phase or step should be run or not.
+
+#### `integrationStepPhases`
+
+The `integrationStepPhases` field is used to define how an integration collects
+data. It is expected that an array of phases that contain "steps" to perform the
+collection of data needs to be passed in. Phases are run in sequence and steps
+are run in parallel. This allows for data from a previous phase to be referenced
+in a later stage.
+
+A `phase` must contain an `id`, `name`, and array of `steps` to perform.
+
+A `step` must contain an `id`, `name`, list of `types` that the step expects to
+generate and an `executionHandler` function that performs the data collection
+work. It is important to provide `types` because that data will be used to
+provide context for the backend synchronization process and determine how
+updates and deletes should be applied.
+
+Optionally, a `step` can contain a `dependsOn` list that references previous
+phases and steps. This field will be used to determine if previous work has
+failed to complete and if the synchronization process should treat the data
+retrieved in the step as a partial dataset. See the [Failure handling](#Failure
+handling) section below for more information.
+
+### How integrations are executed
+
+The `IntegrationInvocationConfig` declaratively defines how an integration runs.
+The CLI that is provided by this package will consume the invocation
+configuration and construct a state machine to execute work in a specific order.
+
+The `invocationValidator` function will run first to ensure the integration is
+capable of interfacing with a given provider. Next, the `getStepStartStates`
+will run to get a list of integration phases and steps that should be executed.
+Finally, the phases in `integrationStepPhases` are then executed in sequence.
+
+### What's in the `IntegrationExecutionContext` and `IntegrationExecutionStepContext`?
+
+The `IntegrationExecutionContext` contains some basic utilities to provide some
+information about the entity data.
+
+It includes a `logger` that can be used to display additional information
```

Comment:
I think the `logger` would be distinct from a mechanism that allows presenting information to the user in a job recording, assuming the current job events presentation we have today. The `logger` would allow for traditional backend info/debug logging.

How about another thing, `jobLog`, with functions that constrain the class of messages? For example, when a list of resources work (like `listResources`), but details are forbidden with a `400` or `403` response (`describeResource`), we could say `jobLog.forbiddenResource({ err, _type, resource: "someResource" })` and the execution environment would `logger` this and also construct a consistent message for the UI (or collect them and provide a summary).

Reply:
As I was writing this, I was p much omitting the actual event logging and my intent was to have the `logger` just be for debugging since we've talked about the issue with people not really knowing what kind of events to log (my writing sucks so that wasn't clear). 

I like the idea of a `jobLog` with a limited interface though. Providing that would allow for devs to provide some additional information about what could have potentially wrong while a step is being executed. Thoughts on perhaps exposing the `logger` under a different name to clarify that information logged with it will not be displayed in the event log? Perhaps `debugLogger` or `internalLogger`? (@philidem help me out here) 

`forbiddenResource` seems like a good function to start with exposing. Any other functions that might be worth adding?

Reply:
I think the only thing I've ever wanted to expose to the job event log from an integration was permissions problems. The `invocationValidator` function just throws certain types of errors which the execution environment catches and sends to the job event log; this has allowed for communicating authentication problems. Authorization problems tend to show up in specific resources, and that cannot be an exception path because it would kill the step. So, I think the only real need is to send authorization/resource access failures.

Perhaps the `logger` could have another function, since it is a type we control, such as `logger.auth()`, and we not only log that to `console` using Bunyan, but we also send a job event? Then we just have the `logger` alone, (`IntegrationLogger`) and document functions to indicate what will be presented to a user and what will not?

We'd like to send `error()` and `auth()` I think to the job event log.

Reply:
I think I dig it. At first I was thinking it was a bit magical but the more I think about it, the more I like having a unified interface.